sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala

```diff
@@ -271,17 +272,23 @@ abstract class BaseDynamicPartitionDataWriter(
 
     val bucketIdStr = bucketId.map(BucketingUtils.bucketIdToString).getOrElse("")
 
-    // This must be in a form that matches our bucketing format. See BucketingUtils.
-    val ext = f"$bucketIdStr.c$fileCounter%03d" +
+    // The prefix and suffix must be in a form that matches our bucketing format.
+    // See BucketingUtils.
+    val prefix = bucketId match {
```

Comment:
nit: bucketId.map(description.bucketSpec.get.bucketFileNamePrefix).getOrElse("")

Reply:
Actually found unit test failure after making the change ([stack trace](https://gist.github.com/c21/e30a2e1dc7017f2c1c206c46a194ac69), [unit test details](https://github.com/c21/spark/runs/3627478552?check_suite_focus=true)). It looks like to me the `description.bucketSpec.get` seems to be eagerly evaluated with `bucketId.map` here.

Changed back to the previously pattern matching.